src/eval.rs

```diff
@@ -1,26 +1,998 @@
+                    BOp::Lte => Ok(Value::Bool(l <= r)),
+                    BOp::Eq => Ok(Value::Bool(l == r)),
+                    BOp::Ne => Ok(Value::Bool(l != r)),
+
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+
+                (Value::Bool(l), Value::Bool(r)) => match op {
+                    BOp::And => Ok(Value::Bool(l && r)),
+                    BOp::Or => Ok(Value::Bool(l || r)),
+
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+                _ => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::UOp { op, inner } => {
+            let left = eval_expr(inner, env, &program)?;
+
+            match left {
+                Value::Num(l, range) => match op {
+                    UOp::Negate => cast(-l, range, Overflow::Wraparound),
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+
+                Value::Bool(b) => match op {
+                    UOp::Not => Ok(Value::Bool(!b)),
+                    _ => Err(RuntimeError::InvalidOperand),
+                },
+                _ => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::Cast {
+            inner,
+            typ,
+            overflow,
+        } => {
+            let cast_t = match typ {
+                Type::BoolT => Err(RuntimeError::TypeError),
+                Type::SymT => Err(RuntimeError::TypeError),
+                Type::NumT(n) => Ok(n.clone()),
+            }?;
+
+            let val = eval_expr(inner, env, &program)?;
+
+            match val {
+                Value::Num(num, range) => match overflow {
+                    Overflow::Fail => cast(num, cast_t, Overflow::Fail),
+                    Overflow::Saturate => cast(num, cast_t, Overflow::Saturate),
+                    Overflow::Wraparound => cast(num, cast_t, Overflow::Wraparound),
+                },
+                // can only cast ints
+                Value::Bool(b) => Err(RuntimeError::TypeError),
+                Value::Sym(s) => Err(RuntimeError::TypeError),
+            }
+        }
+
+        Expr::Match { scrutinee, cases } => {
+            let raw_val = eval_expr(&scrutinee, env, &program)?;
+            for case in cases {
+                let pattern = match case.pattern {
+                    Pattern::Bool(b) => Value::Bool(b),
+                    Pattern::Num(n) => Value::Num(
+                        n,
+                        Range {
+                            start: n,
+                            end: n + 1,
+                        },
+                    ), // TODO: how would we match on Num, not including the range?
+                    Pattern::Sym(s) => Value::Sym(s),
+                    Pattern::Var(id) => env.get(&id).unwrap().clone(),
+                };
+                if raw_val == pattern {
+                    let g = eval_expr(&case.guard, env, &program).unwrap();
```

Comment:
don't have `unwrap` here (or below). Use `?` or directly return the result. Otherwise, you'll crash the interpreter when the program encounters an error.

Reply:
In other words, use `unwrap` only if you know that failure is impossible or irrecoverable from.